agent/rpc/subscribe/subscribe.go

```diff
@@ -0,0 +1,212 @@
+package subscribe
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/hashicorp/go-hclog"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/hashicorp/consul/acl"
+	"github.com/hashicorp/consul/agent/consul/state"
+	"github.com/hashicorp/consul/agent/consul/stream"
+	"github.com/hashicorp/consul/proto/pbservice"
+	"github.com/hashicorp/consul/proto/pbsubscribe"
+)
+
+// Server implements a StateChangeSubscriptionServer for accepting SubscribeRequests,
+// and sending events to the subscription topic.
+type Server struct {
+	Backend Backend
+	Logger  Logger
+}
+
+func NewServer(backend Backend, logger Logger) *Server {
+	return &Server{Backend: backend, Logger: logger}
+}
+
+type Logger interface {
+	Trace(msg string, args ...interface{})
+	With(args ...interface{}) hclog.Logger
+}
+
+var _ pbsubscribe.StateChangeSubscriptionServer = (*Server)(nil)
+
+type Backend interface {
+	ResolveToken(token string) (acl.Authorizer, error)
```

Comment:
Does streaming need to perform ACL token based namespace inference?

In many other places we call the other `ResolveTokenAndDefaultMeta` to perform the `EnterpriseMeta` defaulting based on the token and my guess is that we will want subscriptions to also do this defaulting.

Reply:
I have no idea, but I guess so, if that's what we do for other RPC endpoints. I'll change this interface to `ResolveTokenAndDefaultMeta`

Reply:
I don't fully understand what this does. In what cases would we need to perform ACL token based namespace inference, vs not? The token can have a default namespace, if one isn't set in the request?

The `SubscriptionRequest` doesn't have an EnterpriseMeta.Namespace yet, so if the above is correct, I think this be done when adding the Namespace to SubscriptionRequest, which should be one of the next follow up PRs.